### PR TITLE
added fungtion modules (download models + run)

### PR DESCRIPTION
**Pull request by vagkaratzas**
CPU version of the module

- [ ] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Broadcast software version numbers to `topic: versions` - [See version_topics](https://nf-co.re/blog/2025/version_topics)
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [x] `nf-core modules test <MODULE> --profile docker`
    - [x] `nf-core modules test <MODULE> --profile singularity`
    - [x] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/fungtion/downloadmodels/environment.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+---
+# yaml-language-server: $schema=https://raw.githubusercontent.com/nf-core/modules/master/modules/environment-schema.json
+channels:
```
Comment (mberacochea) [inline modules/nf-core/fungtion/downloadmodels/environment.yml:3]: It would be nice to push this to bioconda

File: modules/nf-core/fungtion/downloadmodels/tests/main.nf.test [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    tag "fungtion"
+    tag "fungtion/downloadmodels"
+
+    test("stub") {
+
+        options "-stub"
+
+        when {
+            process {
+                """
```
Comment (mberacochea) [inline modules/nf-core/fungtion/downloadmodels/tests/main.nf.test:18]: This seems a bit useless

File: modules/nf-core/fungtion/fungtion/main.nf [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    output:
+    tuple val(meta), path("${prefix}/${prefix}.csv")             , emit: predictions
+    tuple val(meta), path("${prefix}/${prefix}_analysis")        , emit: analysis,  optional: true
+    tuple val(meta), path("${prefix}/${prefix}.html")            , emit: html,      optional: true
+    tuple val(meta), path("${prefix}/${prefix}_assets")          , emit: assets,    optional: true
+    tuple val(meta), path("${prefix}/${prefix}_temp_folder")     , emit: temp,      optional: true
+    tuple val(meta), path("${prefix}.log")                       , emit: log
+    tuple val("${task.process}"), val('fungtion'), eval("fungtion --version 2>&1 | head -1"), topic: versions, emit: versions_fungtion
+    tuple val("${task.process}"), val('python'), eval("python --version | sed 's/Python //'"), topic: versions, emit: versions_python
+    tuple val("${task.process}"), val('torch'), eval("python -c 'import torch; print(torch.__version__)'"), topic: versions, emit: versions_torch
```
Comment (mberacochea) [inline modules/nf-core/fungtion/fungtion/main.nf]: Why torch as a separate bit? Is a dependency of `fungtion` (https://github.com/noHup-cc/fungtion/blob/main/pyproject.toml#L23)

File: modules/nf-core/fungtion/fungtion/main.nf [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    def args = task.ext.args ?: ''
+    prefix = task.ext.prefix ?: "${meta.id}"
+    """
+    fungtion \\
+        $args \\
+        --fasta ${fasta} \\
+        --output-dir . \\
+        --prefix ${prefix} \\
+        --pretrain ${pretrain}/esm1b_t33_650M_UR50S.pt \\
+        > ${prefix}.log
```
Comment (mberacochea) [inline modules/nf-core/fungtion/fungtion/main.nf]: This redirect will eat up the error log, so I would only redirect 1> there and leave 2 just to the std to it ends in the .command.err.

File: modules/nf-core/fungtion/fungtion/meta.yml [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+          ontologies: []
+  log:
+    - - meta:
+          type: map
+          description: |
+            Groovy Map containing sample information
+            e.g. `[ id:'sample1' ]`
+      - ${prefix}.log:
+          type: file
+          description: Log file containing the stdout and stderr output of the fungtion run.
```
Comment (mberacochea) [inline modules/nf-core/fungtion/fungtion/meta.yml]: Why are the log an outputs?

**Review by mberacochea — COMMENTED** [review]
Looks good for the most part, I've left some comments/questions.

Also this tool supports GPUs too, shouldn’t we test that one too?

File: modules/nf-core/fungtion/downloadmodels/environment.yml (hunk 1 above)
Comment (vagkaratzas) [reply modules/nf-core/fungtion/downloadmodels/environment.yml:3]: https://github.com/bioconda/bioconda-recipes/pull/66615

File: modules/nf-core/fungtion/downloadmodels/tests/main.nf.test (hunk 2 above)
Comment (vagkaratzas) [reply modules/nf-core/fungtion/downloadmodels/tests/main.nf.test:18]: it is in similar fashion to the `caalm` one..because the `fungtion/fungtion` will download this anyway in the setup step to test..saving some compute by not downloading the models twice

File: modules/nf-core/fungtion/fungtion/meta.yml (hunk 5 above)
Comment (vagkaratzas) [reply modules/nf-core/fungtion/fungtion/meta.yml]: Just for sanity checking that all steps run properly. Looks like this:
```Starting Fungtion prediction run
Input FASTA: protein_mini_with_cazymes.faa
Output directory: ./test
Using ESM-1b weights: models/esm1b_t33_650M_UR50S.pt
Feature extraction device: auto
Extracting ESM-1b features...
ESM-1b features saved to ./test/test_temp_folder/features.csv
Running fungal effector prediction...
Prediction CSV saved to ./test/test.csv
Generating network and tree visualizations...
Visualization manifest saved to test/test_analysis/visualization_manifest.csv
Generating HTML report...
HTML report saved to ./test/test.html
Intermediate files kept in ./test/test_temp_folder
Prediction finished. Results saved to ./test/test.csv
```

File: modules/nf-core/fungtion/fungtion/main.nf (hunk 3 above)
Comment (vagkaratzas) [reply modules/nf-core/fungtion/fungtion/main.nf]: True, not needed here, removing

File: modules/nf-core/fungtion/fungtion/main.nf (hunk 4 above)
Comment (vagkaratzas) [reply modules/nf-core/fungtion/fungtion/main.nf]: This should only redirect the stdout. I'll update to `1>` for clarity though.The meta.yml descr was wrong -> updating that as well.

**Comment by vagkaratzas** [comment]
> Also this tool supports GPUs too, shouldn’t we test that one too?

GPU versions of modules should be different nf-core modules. Should follow these guideliness to create those: https://github.com/nf-core/website/pull/4142

(need to make one for `caalm` as well)

**Comment by mberacochea** [comment]
> > Also this tool supports GPUs too, shouldn’t we test that one too?
> 
> GPU versions of modules should be different nf-core modules. Should follow these guideliness to create those: [nf-core/website#4142](https://github.com/nf-core/website/pull/4142)
> 
> (need to make one for `caalm` as well)

Perfect, I was trying to find the guidelines for those. Then should we set `--device cpu` on this one?

File: modules/nf-core/fungtion/fungtion/meta.yml (hunk 5 above)
Comment (mberacochea) [reply modules/nf-core/fungtion/fungtion/meta.yml]: OK, I would use tee then `| tee ${prefix}.log` to get the output in the log file plus the output and errors in the corresponding .command.out and .command.err

File: modules/nf-core/fungtion/fungtion/main.nf (hunk 4 above)
Comment (mberacochea) [reply modules/nf-core/fungtion/fungtion/main.nf]: yes, I left a suggestion to use use tee so the .command.out is also populated.

File: modules/nf-core/fungtion/downloadmodels/environment.yml (hunk 1 above)
Comment (mberacochea) [reply modules/nf-core/fungtion/downloadmodels/environment.yml:3]: Perfect, I would wait then to merge this one.. well, or just update it afterwards.

**Comment by vagkaratzas** [comment]
> Perfect, I was trying to find the guidelines for those. Then should we set `--device cpu` on this one?

Alright, let's be explicit. If somebody wants to try somethign hacky to make this work with cuda, they should know to come and change that

**Review by mberacochea — APPROVED** [review]
Last bit to adjust is the pipe of the output to the .command.out too with tee.

File: modules/nf-core/fungtion/fungtion/meta.yml (hunk 5 above)
Comment (vagkaratzas) [reply modules/nf-core/fungtion/fungtion/meta.yml]: Added this as well:
```
# Without this, the pipeline exit status can become tee's exit status,
# which can mask a failing fungtion command.
set -o pipefail
```

File: modules/nf-core/fungtion/downloadmodels/environment.yml (hunk 1 above)
Comment (vagkaratzas) [reply modules/nf-core/fungtion/downloadmodels/environment.yml:3]: I found that for caalm it was easier to use the pypi packages. I'll merge this and try to update later, when doing the GPU version
